Major rework: more pythonic API, code rework

Mattwmaster58 opened the pull request:
Here's the things I did (which fixes #37 ):
 - refactored code to 
 - ran black
 - made function docstring
 - added type hints here and there
 - refactored to allow things like a contrib module

What I didn't do (possibly yet):
 - optimize for performance
   - if doing so, I'd like to use numpy (if it exists) to optimize stuff
     - if that's not possible or/and I dont't feel like that, I guess I could perform micro-optimizations I used to make the code more readable if necessary
 - test deliberate failures (reason for decreased coverage)

~~What I need to do:~~
 - [x] update docs

Mattwmaster58 commented:
After thinking about this some more, it probably would make more send to import PIL.Image gracefully and then check with isinstance. Otherwise, how would the user load the code in the first place?

whtsky commented on pixelmatch.py:
```diff
@@ -1,45 +1,105 @@
+
+def pixelmatch(
+    img1: PILImageOrImageSequence,
+    img2: PILImageOrImageSequence,
```
I think the support for PIL should inside a seperated function, or maybe a different import name(say `pixelmatch.contrib.PIL` .

Adding support for PIL directly inside the main `pixelmatch` function might cause some maintenance problem if we want to add support for other libraries ( say opencv or matplotlib ) as well.

Mattwmaster58 replied:
I think you're right on this one. Didn't even consider adding support for other images. The only benefit with this approach would be that you could mix ImageSequence and PILImage objects, together. However, if you are mixing like that I think it's fair to offload the responsibility of that to you.

Mattwmaster58 commented on pyproject.toml:
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "pixelmatch"
-version = "0.1.1"
+version = "1.0.0"
```
Wasn't aware of this nuance. Thanks.

Mattwmaster58 commented on pixelmatch.py:
```diff
@@ -1,45 +1,105 @@
+from typing import Sequence, Union, List, Tuple
+
+
+ImageData = List[int]
```
I think that this type is only used internally and as output, so there's no need to broaden it IMO. I'll need to look into that.

Mattwmaster58 replied:
`Union[int, float]` is superfluous, `float` covers both of those when it comes to Mypy.

Mattwmaster58 commented on pixelmatch.py:
```diff
@@ -1,45 +1,105 @@
+
+
+ImageData = List[int]
+PILImageOrImageArray = Union["Image", Sequence[float]]
```
It's my understanding that it should work if mypy's MYPY_PATH is setup correctly.

whtsky replied:
I tried to run mypy locally but got the `No library stub file for module 'PIL'` error.

Looks like pillow does not support type hints: https://github.com/python-pillow/Pillow/issues/2625 .

Mattwmaster58 replied:
Well, Pillow itself doesn't. But that doesn't mean we can type hint that an object should be an instance of a class. This is moot though as we'll be able to properly type hint it with the crontrib module.

Mattwmaster58 commented on pixelmatch.py:
```diff
@@ -1,45 +1,105 @@
+
+ImageData = List[int]
+PILImageOrImageArray = Union["Image", Sequence[float]]
+RGBTuple = Tuple[int, int, int]
```
I thought about this but opted for a stricter definition so that you don't get something like [1,2,3,4] and your IDE can't warn you because it fits in the type definition of List[int] ~~I think that one could use the new 3.8 exclusive `Literal` type.~~ That wouldn't work either.

Most people messing around with images would be smart enough to know where the error is coming from if they provided an `A` value, and I guess if we broaden this the IDE (correctly) won't complain about `[1,1,1]`.

Mattwmaster58 commented:
Totally get WRT splitting up PRs. I'll try to get a different one out for PIL support

Mattwmaster58 commented:
Alright. I removed all the PIL specific code and moved pixelmatch to `__init__` to lay the groundwork for a contrib module. And type hinted the rest of the code.

whtsky commented on pixelmatch/__init__.py:
```diff
@@ -223,19 +235,31 @@ def rgb2q(r: int, g: int, b: int):
 
 
-def draw_pixel(output, pos: int, r: int, g: int, b: int):
+def draw_pixel(output: MutableImageSequence, pos: int, r: int, g: int, b: int):
```
`MutableImageSequence` is typed as `MutableSequence[float]`: https://github.com/whtsky/pixelmatch-py/pull/38/files#diff-c9a626a30caa55ab4ae697edf9b3e74aR4

But we are writing int into it..

---

Sadly users can't pass a `MutableSeuqnce[int]` here.

![图片](https://user-images.githubusercontent.com/876694/81426217-78b79400-918b-11ea-91cd-24e74ca0acb9.png)

whtsky replied:
I have to say get types hints done right in Python is pretty hard, and that's why I didn't add type hints for image data at first.

Mattwmaster58 replied:
From [PEP 484](https://www.python.org/dev/peps/pep-0484/):

> Rather than requiring that users write import numbers and then use numbers.Float etc., this PEP proposes a straightforward shortcut that is almost as effective: when an argument is annotated as having type float, an argument of type int is acceptable; similar, for an argument annotated as having type complex, arguments of type float or int are acceptable.

It looks like you are using Pyright. This is a known issue that has been fixed, though: https://github.com/microsoft/pyright/issues/260

Maybe your Pyright version is out of date?

whtsky replied:
I'm using VSCode with [Pyright](https://marketplace.visualstudio.com/items?itemName=ms-pyright.pyright) extension v1.1.35

whtsky replied:
I can reproduce this with mypy 0.770.
```python
from typing import MutableSequence

a: MutableSequence[int] = [255, 255, 0]
aa_color: MutableSequence[float] = a
```
![图片](https://user-images.githubusercontent.com/876694/81428525-ea451180-918e-11ea-83cc-13ce02a39be7.png)


( Sorry for the color scheme, I was just too lazy to change them in VSCode)

Mattwmaster58 replied:
lol. On the surface it seems like a bug, but since it exists in both pyright and mypy I'll just work around it.

whtsky replied:
I added mypy in #41 . This should make things a bit easier.

Mattwmaster58 commented:
I had a vague idea that git didn't track empty files/folders, turns out it's only empty folder 😄

whtsky commented on README.md:
```diff
@@ -94,7 +88,7 @@ img_diff.save("diff.png")
 
 ### vnext
 
-- docs: use absolute url for images in READM
```
please don't remove this

whtsky commented on pixelmatch/__init__.py:
```diff
@@ -1,17 +1,54 @@
+    img2: ImageSequence,
+    width: int,
+    height: int,
+    output: MutableImageSequence = None,
```
`Optional[MutableImageSequence]`

Mattwmaster58 commented:
Done

whtsky commented:
https://github.com/Mattwmaster58/pixelmatch-py/runs/657071069?check_suite_focus=true

There are some type errors :(

Mattwmaster58 commented:
Fixed

whtsky commented on test_pixelmatch.py:
```diff
@@ -1,4 +1,6 @@
+import itertools
```
What is itertools for?

whtsky commented on README.md:
```diff
@@ -94,7 +90,7 @@ img_diff.save("diff.png")
 ### vnext
 
-- docs: use absolute url for images in README
+- docs: use absolute url for images in READM
```
Guess you chosed the wrong side when resolving merge conflict

whtsky commented:
> Maybe some pre-commit hooks to run black would be in order 😉

Personally I use editor's auto format: https://github.com/psf/black#editor-integration

Python's [pre-commit](https://github.com/pre-commit/pre-commit) requires manual install, and I have some unpleasant experience with it due to Python venv switching.

whtsky commented:
Merged, thanks!

whtsky commented:
https://pypi.org/project/pixelmatch/0.2.0/

Mattwmaster58 commented:
Thanks for the quick release!

On Fri, May 8, 2020 at 12:33 PM Wu Haotian <notifications@github.com> wrote:

> https://pypi.org/project/pixelmatch/0.2.0/
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/whtsky/pixelmatch-py/pull/38#issuecomment-625955652>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGI56LPJFOMCTBXGB6446HTRQRF7PANCNFSM4M3RAQCA>
> .
>


-- 
*Either I will find a way, or I will make one.** - Phillip Sydney*
